File: 백준/Platinum/10999. 구간 합 구하기 2/구간 합 구하기 2.py
import sys

input = sys.stdin.readline

MAX = 1 << 21


def func(a, b):
    return a + b


class ST:
    def __init__(self):
        self.start = MAX // 2
        self.arr = [0] * MAX
        self.lazy = [0] * MAX

    def build(self):
        for i in range(self.start - 1, 0, -1):
            self.arr[i] = func(self.arr[i * 2], self.arr[i * 2 + 1])

    def propagate(self, node, l, r):
        if self.lazy[node] != 0:
            if node < self.start:
                self.lazy[node * 2] = func(self.lazy[node * 2], self.lazy[node])
                self.lazy[node * 2 + 1] = func(self.lazy[node * 2 + 1], self.lazy[node])
            self.arr[node] += self.lazy[node] * (r-l)
            self.lazy[node] = 0

    def update(self, s, e, k):
        self.update2(s, e, k, 1, 0, self.start)

    def update2(self, s, e, k, node, l, r):
        self.propagate(node, l, r)
        if e <= l or r <= s:
            return
        if s <= l and r <= e:
            self.lazy[node] = func(self.lazy[node], k)
            self.propagate(node, l, r)
            return
        mid = (l + r) // 2
        self.update2(s, e, k, node * 2, l, mid)
        self.update2(s, e, k, node * 2 + 1, mid, r)
        self.arr[node] = func(self.arr[node * 2], self.arr[node * 2 + 1])

    def query(self, s, e):
        return self.query2(s, e, 1, 0, self.start)

    def query2(self, s, e, node, l, r):
        self.propagate(node, l, r)
        if e <= l or r <= s:
            return 0
        if s <= l and r <= e:
            return self.arr[node]
        mid = (l + r) // 2
        return func(self.query2(s, e, node * 2, l, mid), self.query2(s, e, node * 2 + 1, mid, r))


n, m, k = map(int, input().split())

tree = ST()
for i in range(n):
    tree.arr[tree.start + i] = int(input())
tree.build()

for _ in range(k + m):
    op = list(map(int, input().split()))
    if op[0] == 1:
        a, b, c = op[1], op[2], op[3]
        tree.update(a - 1, b, c)
    else:
        a, b = op[1], op[2]
        print(tree.query(a - 1, b))

File: 백준/Platinum/10999. 구간 합 구하기 2/구간 합 구하기 2.cc
#include <bits/stdc++.h>
#pragma GCC optimize("O3")
#pragma GCC optimize("Ofast")
#pragma GCC optimize("unroll-loops")
#define endl '\n'
const int oOvOo = 0;
using ll = long long;
using namespace std;

class LST
{
public:
    int n;
    vector<ll> arr;
    vector<ll> lazy;

    LST(const vector<ll> &arr) : lazy(4 * arr.size())
    {

        n = arr.size();
        this->arr = vector<ll>(4 * arr.size());
        init(arr, 0, n - 1, 1);
    }

    auto func(ll a, ll b)
    {
        return a + b;
    }

    ll init(const vector<ll> &arr, int left, int right, int node)
    {
        if (left == right)
            return this->arr[node] = arr[left];
        int mid = (left + right) / 2;
        ll l = init(arr, left, mid, node * 2);
        ll r = init(arr, mid + 1, right, node * 2 + 1);
        return this->arr[node] = func(l, r);
    }

    void propagate(int node, int nodeLeft, int nodeRight)
    {
        if (lazy[node])
        {
            if (nodeLeft != nodeRight)
            {
                lazy[node * 2] = func(lazy[node * 2], lazy[node]);
                lazy[node * 2 + 1] = func(lazy[node * 2 + 1], lazy[node]);
            }
            arr[node] = func(arr[node], lazy[node] * (nodeRight - nodeLeft + 1));
            lazy[node] = 0;
        }
    }

    ll query(int left, int right)
    {
        return query(left, right, 1, 0, n - 1);
    }
    ll query(int left, int right, int node, int nodeLeft, int nodeRight)
    {
        propagate(node, nodeLeft, nodeRight);
        if (right < nodeLeft || nodeRight < left)
            return 0;
        if (left <= nodeLeft && nodeRight <= right)
            return arr[node];
        int mid = (nodeLeft + nodeRight) / 2;
        return func(query(left, right, node * 2, nodeLeft, mid), query(left, right, node * 2 + 1, mid + 1, nodeRight));
    }

    ll update(int left, int right, ll newValue)
    {
        return update(left, right, newValue, 1, 0, n - 1);
    }

    ll update(int left, int right, ll newValue, int node, int nodeLeft, int nodeRight)
    {
        propagate(node, nodeLeft, nodeRight);
        if (right < nodeLeft || nodeRight < left)
            return arr[node];
        if (left <= nodeLeft && nodeRight <= right)
        {
            lazy[node] = func(lazy[node], newValue);
            propagate(node, nodeLeft, nodeRight);
            return arr[node];
        }
        int mid = (nodeLeft + nodeRight) / 2;
        return arr[node] = func(update(left, right, newValue, node * 2, nodeLeft, mid), update(left, right, newValue, node * 2 + 1, mid + 1, nodeRight));
    }
};

int main()
{
    std::ios_base::sync_with_stdio(false);
    std::cin.tie(nullptr);
    std::cout.tie(nullptr);

    int n, m, k;
    cin >> n >> m >> k;

    vector<ll> arr(n);

    for (auto &i : arr)
        cin >> i;
    LST lst(arr);
    for (int i = 0; i < m + k; i++)
    {
        int a;
        cin >> a;
        if (a == 1)
        {
            int b, c;
            ll d;
            cin >> b >> c >> d;
            lst.update(b-1, c - 1, d);
        }
        else
        {
            int b, c;
            cin >> b >> c;
            cout << lst.query(b - 1, c-1) << endl;
        }
    }

    return oOvOo;
}